### PR TITLE
Move warnings after validity check

### PR DESCRIPTION
**Pull request by octosteve**
Move warning into the environment.

### CODE REVIEW

File: lib/learn_open/environments/ide_environment.rb [hunk 1]
```diff
@@ -2,43 +2,55 @@ module LearnOpen
   module Environments
     class IDEEnvironment < BaseEnvironment
       def open_readme(lesson)
-        if valid?(lesson)
+        when_valid(lesson) do
```
Comment (talum) [inline lib/learn_open/environments/ide_environment.rb:5]: This sounds super dope.

File: lib/learn_open/environments/ide_environment.rb [hunk 2]
```diff
@@ -2,43 +2,55 @@ module LearnOpen
...
-          start_file_backup(lesson, location)
-          install_dependencies(lesson, location)
-          notify_of_completion
-          open_shell
+        when_valid(lesson) do
+          warn_if_necessary(lesson)
+          case lesson
+          when LearnOpen::Lessons::IosLesson
+            super
+          when -> (l) { valid?(l) }
```
Comment (talum) [inline lib/learn_open/environments/ide_environment.rb:26]: Ah, I still think this proc/lambda in a case statement is crazy, but here is a [post](http://blog.honeybadger.io/rubys-case-statement-advanced-techniques/) for anyone who is curious about why/how this works.

**Review by talum — COMMENTED** [review]
👍

File: lib/learn_open/lessons/base_lesson.rb [hunk 3]
```diff
@@ -45,16 +45,6 @@ def readme?
         !lab?
       end
 
-      def warn_if_necessary
```
Comment (ktravers) [inline lib/learn_open/lessons/base_lesson.rb:48]: @StevenNunez cool, i like that it's the environment's responsibility to know whether or not to throw this warning, not the lesson's

**Pull request merged**